src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetScopeManager.cs

```diff
@@ -38,39 +39,52 @@ public Scope Active
 
         public Scope Activate(Span span, bool finishOnClose)
         {
-            var activeScope = Active;
-            var scope = new Scope(activeScope, span, this, finishOnClose);
-            var scopeOpenedArgs = new ScopeEventArgs(scope);
+            var newParent = Active;
+            var scope = new Scope(newParent, span, this, finishOnClose);
+            var scopeOpenedArgs = new SpanEventArgs(span);
+
+            SpanOpened?.Invoke(this, scopeOpenedArgs);
 
-            ScopeOpened?.Invoke(this, scopeOpenedArgs);
             SetScope(scope);
 
-            if (activeScope != null)
+            if (newParent != null)
             {
-                ScopeDeactivated?.Invoke(this, new ScopeEventArgs(activeScope));
+                SpanDeactivated?.Invoke(this, new SpanEventArgs(newParent.Span));
             }
 
-            ScopeActivated?.Invoke(this, scopeOpenedArgs);
+            SpanActivated?.Invoke(this, scopeOpenedArgs);
+
             return scope;
         }
 
         public void Close(Scope scope)
         {
             var current = Active;
-            if (current != null && current == scope)
+            var isRootSpan = scope.Parent == null;
+
+            if (current == null || current != scope)
             {
-                // if the scope that was just closed was the active scope,
-                // set its parent as the new active scope
-                SetScope(current.Parent);
-                ScopeDeactivated?.Invoke(this, new ScopeEventArgs(current));
+                Log.Error($"Tried to close a non-current scope. TraceId: {scope?.Span?.TraceId}, SpanId: {scope?.Span?.SpanId}");
```

Comment:
We [already removed a log message here](https://github.com/DataDog/dd-trace-dotnet/pull/325) because this can be a normal occurrence. Non-active scopes can be closed since spans don't always finish in reverse order in which they started.